### PR TITLE
2159 add bottom navigation bar

### LINKED ISSUE #2159: Add navigation component and a bottom navigation bar to a new navigation activity
**Is your feature request related to a problem? Please describe.**

The new navigation ticket #2112 will re-write many activities into fragments. This should be done piecemeal to reduce mistakes. This second ticket (the first being #2147) will implement the navigation component, add the bottom navigation bar and its destinations as fragments. 

**Describe the solution you'd like**

Using the [navigation component](https://developer.android.com/guide/navigation/navigation-getting-started). This will also branch the unmerged #2147 branch. 

**Additional context**

This is a child ticket to #2112 .

### PR DESCRIPTION
**Pull request by Frans-Lukas**
Fixes #2159 

1. Navigation component is now used to navigate between `reader`, `library` and `download`.  Uses:
    1. res/navigation/kiwix_nav_graph.xml which explains the destinations and their relationships. 
    2. res/layout/activity_new_navigation.xml which contains the NavigationHostFragment that allows for destination switching (in the fragment). It also contains the `BottomNavigatioinView`.
    3. build.gradle plugin safeargs and androidx.navigation imports. 
2. Added three fragment destinations to app.org.kiwix.kiwixmobile.nav.destination
	1. library OnlineLibraryFragment
		- Copy of LibraryFragment
	2. library LocalLibraryFragment
		- Copy of ZimFileSelectFragment
	3. reader ReaderFragment
		- Rewrite of old KiwixMainActivity
3. Each fragment implements its own separate toolbar.

Copying LibraryFragment and ZimFileSelectFragment to new fragments worked nicely. You mentioned that they share viewmodel. How should I proceed with 'fixing' that, if we think it is necessary.

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -84,4 +85,13 @@ dependencies {
   implementation(Libs.squidb_annotations)
   implementation(Libs.ink_page_indicator)
   add("kapt", Libs.squidb_processor)
+
+  // Navigation component
+  val navVersion = "2.3.0-rc01"
```
Comment (macgills) [inline app/build.gradle.kts:90]: We will want to put this in Libs, if you go to Libs.kt you will find the command to update that file, it should automatically add this new one and then you replace your usage here with `Libs.navigation` or whatever it will call it.
Sometimes the command shuffles the file around a bit but I think it has been behaving recently, best to do this as a standalone commit

**Comment by macgills** [comment]
Little bit of feedback:

The OnlineLibraryFragment is partially hidden under the toolbar meaning you can't see the refresh/loading
The bottomnav view doesn't hide when scrolling content in Reader
Action mode does not get cancelled when moving between the top level destinations
What is the guidelines around the enter exit animations of these fragments? Without activity transitions it looks a bit harsh
The backbutton can navigate between the top level destinations, ie by pressing back I can go from reader to Library

**Comment by Frans-Lukas** [comment]
> ...
> The backbutton can navigate between the top level destinations, ie by pressing back I can go from reader to Library

The default behavior is to either exit (if on the `home` destination) or return to the `home` destination (if on any other destination). This is compared to Play store and Play books. Google photos exits the application on back press. I can think of two alternatives:

1. Put the home destination as the left-most destination in the bottom nav bar, this is how play books and store does it and I think it is more natural than having it in the middle.
2. Do it like google photos. I.e. no navigation between destinations using back button but only exiting if on a top-level page in a destination. 

>...
>What is the guidelines around the enter exit animations of these fragments? Without activity transitions it looks a bit harsh
>...

The navigation component provides one, and only one, animation between transitions: Crossfade. I can't tell if it works for us though. It seems unnecessarily difficult to change transition animation but if I can get the default crossfade to work I think it should be fine?  Cross-fade is also what is recommended by [material design.](https://material.io/components/bottom-navigation#behavior)

**Comment by macgills** [comment]
>Do it like google photos. I.e. no navigation between destinations using back button but only exiting if on a top-level page in a destination.

I prefer this option mostly for the mental model I have for navigationbars.


Try and get crossfade working, you can slow down animations in developer settings on your device and that might help diagnose their existence

File: app/src/main/java/org/kiwix/kiwixmobile/main/KiwixNewNavigationActivity.kt [hunk 2]
```diff
@@ -82,7 +82,7 @@ class KiwixNewNavigationActivity : CoreMainActivity() {
   override fun onBackPressed() {
     supportFragmentManager.fragments.filterIsInstance<BaseFragmentActivityExtensions>().forEach {
       if (it.onBackPressed(this) == BaseFragmentActivityExtensions.Super.ShouldCall) {
-        super.onBackPressed()
+        finish()
```
Comment (Frans-Lukas) [inline app/src/main/java/org/kiwix/kiwixmobile/main/KiwixNewNavigationActivity.kt:85]: I want to check if the current fragment is a top level destination and otherwise call super but the only reliable way I found is to compare the current destination ID with all top level destinations which seems silly. https://stackoverflow.com/questions/52111516/in-android-navigation-architecture-how-can-i-check-if-current-fragment-is-the-l

File: app/src/main/java/org/kiwix/kiwixmobile/main/KiwixNewNavigationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_new_navigation)
+    val navView: BottomNavigationView = findViewById(R.id.nav_view)
+
+    navController = findNavController(R.id.nav_host_fragment)
+    navController.addOnDestinationChangedListener(finishActionModeOnDestinationChange)
+
+    appBarConfiguration = AppBarConfiguration(
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/main/KiwixNewNavigationActivity.kt:58]: this can be 1 line

File: app/src/main/java/org/kiwix/kiwixmobile/main/NavigationHostFragment.kt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    return result
+  }
+
+  override fun onNewIntent(
+    intent: Intent,
+    activity: AppCompatActivity
+  ): Super {
+    var result = ShouldCall
+    childFragmentManager.fragments.filterIsInstance<BaseFragmentActivityExtensions>().forEach {
+      result = it.onNewIntent(intent, activity)
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/main/NavigationHostFragment.kt:90]: so this only stores the last result, should there be a type of and-ing here? Or is expected behaviour that previous results are ignored?

**Comment by macgills** [comment]
I didn't get to review the code as much as I would like so I am afraid we have to wait until Monday to go much further.

Wow, completely changes the app doesn't it, just great. It is making me hate our action bars a lot, so clunky and just repeating what the nav bar is saying, would be so much prettier and cleaner without them and the crossfade wouldn't look like it flashes white at the top. I am thinking a lot about [fabs that emit actions](https://material.io/components/buttons-floating-action-button#types-of-transitions) but this is just me being overexcited about the new toy.

The most worrying thing right now is it can chug a bit switching destination which makes me think there is a memory leak of some description. I have in fact replicated the new navigationActivity getting killed due to memory constraints I imagine? Logcat isn't giving me any good hints. This might be a time to break out the memory profiler in android studio

File: app/src/main/java/org/kiwix/kiwixmobile/main/NavigationHostFragment.kt (hunk 4 above)
Comment (Frans-Lukas) [reply app/src/main/java/org/kiwix/kiwixmobile/main/NavigationHostFragment.kt]: ooh, good question. I am thinking it should only call super if all children agree on it calling super. (and-ing here)

File: app/src/main/res/layout/fragment_destination_library.xml [hunk 5]
```diff
@@ -16,33 +16,47 @@
...
+    app:layout_constraintStart_toStartOf="parent"
+    app:layout_constraintTop_toBottomOf="@+id/file_management_no_files" />
 
   <androidx.swiperefreshlayout.widget.SwipeRefreshLayout
     android:id="@+id/zim_swiperefresh"
     android:layout_width="match_parent"
-    android:layout_height="match_parent">
+    android:layout_height="match_parent"
+    tools:layout_editor_absoluteX="16dp"
+    tools:layout_editor_absoluteY="-142dp">
```
Comment (macgills) [inline app/src/main/res/layout/fragment_destination_library.xml:52]: this might have been an accident

Comment (Frans-Lukas) [reply app/src/main/res/layout/fragment_destination_library.xml:52]: nice catch!

File: app/src/main/java/org/kiwix/kiwixmobile/main/KiwixNewNavigationActivity.kt [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+  private val finishActionModeOnDestinationChange =
+    NavController.OnDestinationChangedListener { controller, destination, arguments ->
+      actionMode?.finish()
+    }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_new_navigation)
+    val navView: BottomNavigationView = findViewById(R.id.nav_view)
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/main/KiwixNewNavigationActivity.kt:53]: shouldn't need findViewById unless this is across the module boundary which it probably shouldn't be

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/fileselect_view/effects/OpenFileWithNavigation.kt [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  SideEffect<Unit> {
+
+  override fun invokeWith(activity: AppCompatActivity) {
+    val file = bookOnDisk.file
+    if (!file.canRead()) {
+      activity.toast(R.string.error_file_not_found)
+    } else {
+      val action = actionNavigationLibraryToNavigationReader()
+      action.zimFileUri = file.toUri().toString()
+      activity.findNavController(R.id.nav_host_fragment).navigate(action)
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/zim_manager/fileselect_view/effects/OpenFileWithNavigation.kt:40]: Will we have multiple NavControllers?

I think this and other code can go to an extension function 
```
activity.navigate(
  actionNavigationLibraryToNavigationReader().apply{ zimFileUri = file.toUri().toString() }
)
```

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.kt [hunk 8]
```diff
@@ -61,7 +63,7 @@ import org.kiwix.kiwixmobile.zim_manager.library_view.adapter.LibraryListItem
 import org.kiwix.kiwixmobile.zim_manager.library_view.adapter.LibraryListItem.BookItem
 import javax.inject.Inject
 
-class LibraryFragment : BaseFragment() {
+open class LibraryFragment : BaseFragment(), BaseFragmentActivityExtensions {
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.kt:66]: shouldn't this interface only be on OnlineLibraryFragment?

File: app/src/main/res/anim/nav_default_enter_anim.xml [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  ~ GNU General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU General Public License
+  ~ along with this program. If not, see <http://www.gnu.org/licenses/>.
+  ~
+  -->
+
+<set xmlns:android="http://schemas.android.com/apk/res/android">
```
Comment (macgills) [inline app/src/main/res/anim/nav_default_enter_anim.xml:19]: My find usages says these animations are unused

File: app/src/main/res/layout/activity_new_navigation.xml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  ~ GNU General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU General Public License
+  ~ along with this program. If not, see <http://www.gnu.org/licenses/>.
+  ~
+  -->
+
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (macgills) [inline app/src/main/res/layout/activity_new_navigation.xml:19]: How about Constraintlayout here instead, everytime I see weightsum I get irrationally angry

File: build.gradle.kts [hunk 11]
```diff
@@ -6,6 +6,10 @@ buildscript {
   dependencies {
     classpath(Libs.com_android_tools_build_gradle)
     classpath(Libs.kotlin_gradle_plugin)
+
+    val navVersion = "2.3.0-rc01"
+    classpath("androidx.navigation:navigation-safe-args-gradle-plugin:$navVersion")
```
Comment (macgills) [inline build.gradle.kts:11]: This should be in Libs

File: core/src/main/res/values/styles.xml [hunk 12]
```diff
@@ -60,6 +60,13 @@
     <item name="cardBackgroundColor">@color/card_view_background</item>
   </style>
 
+  <style name="ThemeOverlay.App.BottomNavigationView" parent="Widget.MaterialComponents.BottomNavigationView">
```
Comment (macgills) [inline core/src/main/res/values/styles.xml:63]: It isn't really a ThemeOverlay when you inherit from a parent, this can just follow our widget naming convention

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/fileselect_view/effects/OpenFileWithNavigation.kt (hunk 7 above)
Comment (Frans-Lukas) [reply app/src/main/java/org/kiwix/kiwixmobile/zim_manager/fileselect_view/effects/OpenFileWithNavigation.kt:40]: There will be multiple navigation actions from the library view and having them as extension functions will be nice!

File: app/src/main/res/anim/nav_default_enter_anim.xml (hunk 9 above)
Comment (Frans-Lukas) [reply app/src/main/res/anim/nav_default_enter_anim.xml:19]: They are not referenced anywhere but overrides the default animations for navigations [1]. Not the slickest solution but I couldn't find anything more clean and simple.

[1] https://stackoverflow.com/questions/50482095/how-do-i-define-default-animations-for-navigation-actions

File: app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.kt (hunk 8 above)
Comment (Frans-Lukas) [reply app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.kt:66]: It uses the `override onBackPressed` implementation here. Why should this fragment not implement the BaseFragmentActivityExtensions?

Comment (macgills) [reply app/src/main/java/org/kiwix/kiwixmobile/zim_manager/library_view/LibraryFragment.kt:66]: I didn't think it was reporting this to anyone because the LibraryFragment specifically is only used in Old navigation from ZimManageActivity

File: app/src/main/res/anim/nav_default_enter_anim.xml (hunk 9 above)
Comment (macgills) [reply app/src/main/res/anim/nav_default_enter_anim.xml:19]: I don't like the resource overriding that much but I guess it will have to do. I guess as long as they are not reported unused by lint?

File: app/src/main/java/org/kiwix/kiwixmobile/ActivityExtensions.kt [hunk 13]
```diff
@@ -31,6 +33,9 @@ private val BaseActivity.kiwixComponent: KiwixComponent
 
 private fun BaseActivity.kiwixApp() = applicationContext as? KiwixApp ?: application as? KiwixApp
 
+fun BaseActivity.navigate(action: NavDirections) =
```
Comment (macgills) [inline app/src/main/java/org/kiwix/kiwixmobile/ActivityExtensions.kt:36]: this can be AppCompatActivity, to avoid the casting at usage sites

File: build.gradle.kts [hunk 14]
```diff
@@ -6,9 +6,7 @@ buildscript {
   dependencies {
     classpath(Libs.com_android_tools_build_gradle)
     classpath(Libs.kotlin_gradle_plugin)
-
-    val navVersion = "2.3.0-rc01"
```
Comment (macgills) [inline build.gradle.kts]: 2.3.0 is out so you can update this

Comment (Frans-Lukas) [reply build.gradle.kts]: Yay!
